feat: Allow passing arbitrary Tensorizer serialization and deserialization kwargs; update docs

sangstar opened the pull request:
Feature adding the following to the [sangstar/tensorizer-aws-update-and-any-kwargs](https://github.com/coreweave/vllm/tree/sangstar/tensorizer-aws-update-and-any-kwargs) branch, which aims to add:

- Exposing arbitrary kwargs to Tensorizer from vLLM
- Harmonizing Tensorizer's AWS-style boto3 support with vLLM's
- Updating the docs to reflect these changes, greatly expanding upon the intended usage pattern so users can read it and immediately know how to get started

This is the kwargs-update portion. This PR adds:
- Exposing arbitrary kwargs to `TensorSerializer` and `TensorDeserializer`, adding tests to confirm functionality
- Update docs thoroughly to reflect straight-forward usage, including the new changes. The docs are not finished yet, as they'll be updated with the AWS-update feature to be included in the eventual PR for the aforementioned branch.

The changes herein are succinctly described in a snippet to the updated docs, describing how to load a model using Tensorizer with the updated arbitrary kwargs without too much pain in constructing the JSON string. The kwargs here are just throwaways demonstrating the support.

```bash
#!/bin/sh

MODEL_LOADER_EXTRA_CONFIG='{
  "tensorizer_uri": "s3://my-bucket/vllm/facebook/opt-125m/v1/model.tensors",
  "stream_kwargs": {"force_http": false},
  "deserialization_kwargs": {"verify_hash": true, "num_readers": 8}
}'

vllm serve facebook/opt-125m \
  --load-format=tensorizer \
  --model-loader-extra-config="$MODEL_LOADER_EXTRA_CONFIG"
```

_Please hold off on any requests for changes relating to formatting. This will all be done in a later stage with vLLM's formatter._

sangstar committed:
tests: Add support for unit testing within the worker processes
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Add a `._debug_ctx`, arbitrary serializer and deserializer kwargs
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Add `--serialization-kwargs` support to `tensorize_vllm_model.py` and add to test
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Rm pointless `._debug_ctx` attribute
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
tests: Rm `_debug` in initializer
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Add test to confirm `deserialization_kwargs` passed to `TensorDeserializer`
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Add `stream_kwargs` and add test for it
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
tests: Update entrypoint test to include new kwargs, add other test
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
docs: Update md on using Tensorizer with vLLM
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
tests: Print outputs from subprocesses for test
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
docs: Fix link
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
docs: Add more updates to md
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
docs: Clarify model tensors location
Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
fix: Allow parsing JSON strings with newlines

Adjusts the regex string in `arg_utils.parse_type` to allow
for newlines within the JSON string

Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
tests: Confirm server starts successfully for `vllm serve` test

Signed-off-by: Sanger Steel <sangersteel@gmail.com>

Eta0 commented on docs/models/extensions/tensorizer.md:
```diff
@@ -8,9 +8,97 @@ vLLM model tensors that have been serialized to disk, an HTTP/HTTPS endpoint, or
 at runtime extremely quickly directly to the GPU, resulting in significantly
 shorter Pod startup times and CPU memory usage. Tensor encryption is also supported.
```
This wasn't added as part of this PR, but this should be rephrased while you're at it.
```suggestion
lower Pod startup times and CPU memory usage. Tensor encryption is also supported.
```

Eta0 commented on docs/models/extensions/tensorizer.md:
```diff
@@ -8,9 +8,97 @@ vLLM model tensors that have been serialized to disk, an HTTP/HTTPS endpoint, or
+The core frontend object of note integrating Tensorizer is 
+`TensorizerConfig`, defined [here](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/model_loader/tensorizer.py#L135-L214)
+It's a config object holding important state and passed to any serialization 
+or deserialization operation. When loading with Tensorizer using the vLLM 
```
Suggesting to change this mainly for succinctness but also so that it doesn't link to a specific line number range on the main branch, since those line numbers will quickly become inaccurate. Also, it doesn't look like that class is actually stateful (i.e. it looks basically immutable), so I'm suggesting dropping the reference to "state" here. If that's not accurate, you can add that part back.
```suggestion
The `TensorizerConfig` class is used to customize Tensorizer's behaviour,
defined in [vllm/model_executor/model_loader/tensorizer.py](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/model_loader/tensorizer.py).
It is passed to any serialization or deserialization operation.
When loading with Tensorizer using the vLLM 
```

Eta0 commented on docs/models/extensions/tensorizer.md:
```diff
@@ -8,9 +8,97 @@ vLLM model tensors that have been serialized to disk, an HTTP/HTTPS endpoint, or
+or deserialization operation. When loading with Tensorizer using the vLLM 
+library rather than through a model-serving entrypoint, it gets passed to 
+the `LLM` entrypoint class directly. Here's an example of loading a model
+saved at `"s3://my-bucket/vllm/facebook/opt-125m/v1/model.tensors"`.
```
```suggestion
saved at `"s3://my-bucket/vllm/facebook/opt-125m/v1/model.tensors"`:
```

Eta0 commented on docs/models/extensions/tensorizer.md:
```diff
@@ -8,9 +8,97 @@ vLLM model tensors that have been serialized to disk, an HTTP/HTTPS endpoint, or
+model_ref = "facebook/opt-125m"
+tensorizer_config = TensorizerConfig(
+        tensorizer_uri=path_to_tensors,
+    )
```
Indentation issue
```suggestion
    tensorizer_uri=path_to_tensors
)
```

Eta0 commented on docs/models/extensions/tensorizer.md:
```diff
@@ -8,9 +8,97 @@ vLLM model tensors that have been serialized to disk, an HTTP/HTTPS endpoint, or
+
+But that code won't work unless you actually have your serialized model 
+tensors `model.tensors`, so let's walk through a basic example of serializing 
+`facebook/opt-125m` using the example script, and then loading it for inference.
```
Nitpick copyediting for professionalism
```suggestion
However, the above code will not function until you have successfully
serialized the model tensors with Tensorizer to get the `model.tensors`
file shown. The following section walks through an end-to-end example
of serializing `facebook/opt-125m` with the example script,
and then loading it for inference.
```

Eta0 commented on docs/models/extensions/tensorizer.md:
```diff
@@ -8,9 +8,97 @@ vLLM model tensors that have been serialized to disk, an HTTP/HTTPS endpoint, or
+initializer, as if you were instantiating it like this, for instance:
+`TensorizerConfig(**{"tensorizer_uri": "foo"})`. Here's an example bash
+script that neatly lays out the JSON string before passing it to
+`--model-loader-extra-config`:
```
Clarity
```suggestion
Once the model is serialized where you want it, you must specify the
following additional command line parameters to `vllm serve` to load the model:
- `--load-format=tensorizer`
- `--model-loader-extra-config`
  - This expects a JSON string specifying keyword arguments for
    the `TensorizerConfig` object
  - For example, `--model-loader-extra-config='{"tensorizer_uri": "foo"}'`
    would be equivalent to `TensorizerConfig(**{"tensorizer_uri": "foo"})`
The shell snippet below shows a typical invocation of `vllm serve`
with these parameters specified:
```

Eta0 commented on docs/models/extensions/tensorizer.md:
```diff
@@ -8,9 +8,97 @@ vLLM model tensors that have been serialized to disk, an HTTP/HTTPS endpoint, or
+}
+EOF
+
+MODEL_LOADER_EXTRA_CONFIG=$(echo "$JSON" | tr -d '\n')
```
Reduce unnecessary complexity here. Also, booleans were being specified as strings like `"False"` instead of proper JSON booleans. If `"False"` is deserialized with `json.load`, it will end up as a truthy value in Python, which isn't what you want here.

Please also update this in the PR description.
```suggestion
MODEL_LOADER_EXTRA_CONFIG='{
  "tensorizer_uri": "s3://my-bucket/vllm/facebook/opt-125m/v1/model.tensors",
  "stream_kwargs": {"force_http": false},
  "deserialization_kwargs": {"verify_hash": true, "num_readers": 8}
}'
```

Eta0 commented on docs/models/extensions/tensorizer.md:
```diff
@@ -8,9 +8,97 @@ vLLM model tensors that have been serialized to disk, an HTTP/HTTPS endpoint, or
+Note in this case, if the directory to the model artifacts at
+`s3://my-bucket/vllm/facebook/opt-125m/v1/` doesn't at least have a `config.
+json` file, you'll want to pass `facebook/opt-125m` as the model tag like
+it was done in the example script above. In our example, we just added a
```
Can you rephrase this slightly for clarity? Maybe something like "the `facebook/opt-125m` positional argument in the above example is required when there is no `config.json` file at [...]"

sangstar replied:
Yes -- I'll update it to account for the changes in the sister PR as well.

Eta0 commented on examples/others/tensorize_vllm_model.py:
```diff
@@ -170,6 +170,14 @@ def parse_args():
+        required=False,
+        help=("A JSON string containing additional keyword arguments that "
+              "will be passed to Tensorizer's `TensorSerializer` during "
+              "serialization."))
```
Nitpick: concision
```suggestion
        help=("A JSON string containing additional keyword arguments to "
              "pass to Tensorizer's `TensorSerializer` during "
              "serialization."))
```

Eta0 commented on examples/others/tensorize_vllm_model.py:
```diff
@@ -295,6 +303,12 @@ def deserialize():
+            serialization_kwargs = json.loads(args.serialization_kwargs)
+            tensorizer_config.serialization_kwargs = serialization_kwargs
+            print("Found serialization kwargs: ", serialization_kwargs)
+
```
Seems to be a stray debug print(?)
```suggestion
```

Eta0 commented on examples/others/tensorize_vllm_model.py:
```diff
@@ -295,6 +303,12 @@ def deserialize():
             **credentials)
 
+        if args.serialization_kwargs:
+            serialization_kwargs = json.loads(args.serialization_kwargs)
```
This parsing/type conversion should be handled at the `ArgumentParser` level by making the `type=` parameter in `serialize_parser.add_argument()` point to a function that calls `json.loads` and then verifies that it got a `dict` as a result (since `json.loads` can also produce things like integers), and that raises a `ValueError` with a meaningful error message if it can't be parsed.

sangstar replied:
Implemented [here](https://github.com/coreweave/vllm/pull/7/commits/2cefde17d3c6e1d812bb6d2084a2478e5061152e).

Eta0 commented on tests/lora/test_llama_tp.py:
```diff
@@ -210,7 +211,8 @@ def test_tp2_serialize_and_deserialize_lora(tmp_path, sql_lora_files,
             str(tp_size), "serialize", "--serialized-directory",
-            str(tmp_path), "--suffix", suffix
+            str(tmp_path), "--suffix", suffix, "--serialization-kwargs",
+            json.dumps({"limit_cpu_concurrency": 4})
```
Seems like overkill
```suggestion
            '{"limit_cpu_concurrency": 4}'
```

Eta0 commented on tests/tensorizer_loader/conftest.py:
```diff
@@ -1,10 +1,18 @@
 
+@pytest.fixture(autouse=True)
+def allow_insecure_serialization():
+    os.environ["VLLM_ALLOW_INSECURE_SERIALIZATION"] = "1"
```
A fixture should clean up after itself afterwards—[the `pytest` documentation](https://docs.pytest.org/en/stable/how-to/monkeypatch.html#monkeypatching-environment-variables) explains how to write a fixture that temporarily changes an environment variable in this way.

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -280,3 +292,196 @@ def test_vllm_tensorized_model_has_same_outputs(vllm_runner, tmp_path):
+                              serialization_kwargs=serialization_params)
+    llm = LLM(
+        model=model_ref,
+        device="cuda",
```
Is `cuda` not the default? Might this test also be relevant in a non-CUDA context if not for writing this in here? Not the biggest deal either way since we mainly care about CUDA.

sangstar replied:
It's the default. There's no real reason have to have this device specification here.

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -280,3 +292,196 @@ def test_vllm_tensorized_model_has_same_outputs(vllm_runner, tmp_path):
+
+        tensorizer.serialization.TensorSerializer.__init__ = tensorizer_serializer_wrapper
+
+        tensorizer_config = TensorizerConfig(**kwargs.get("tensorizer_config"))
```
`.get` isn't helpful here since if it returned `None` it would just end up giving a more confusing error anyway
```suggestion
        tensorizer_config = TensorizerConfig(**kwargs["tensorizer_config"])
```

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -280,3 +292,196 @@ def test_vllm_tensorized_model_has_same_outputs(vllm_runner, tmp_path):
+        tensorizer.serialization.TensorSerializer.__init__ = tensorizer_serializer_wrapper
+
+        tensorizer_config = TensorizerConfig(**kwargs.get("tensorizer_config"))
+        assert tensorizer_config is not None
```
Pretty sure it's impossible for an object instantiation like this to give you back `None` without substantial hackery
```suggestion
```

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -280,3 +292,196 @@ def test_vllm_tensorized_model_has_same_outputs(vllm_runner, tmp_path):
+
+        tensorizer_config = TensorizerConfig(**kwargs.get("tensorizer_config"))
+        assert tensorizer_config is not None
+        self.save_tensorized_model(
+            tensorizer_config=tensorizer_config, )
```
Silly spacing. You can ignore this if this will be resolved by an automatic formatter pass.
```suggestion
        self.save_tensorized_model(tensorizer_config=tensorizer_config)
```

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -280,3 +292,196 @@ def test_vllm_tensorized_model_has_same_outputs(vllm_runner, tmp_path):
+    # lazy_load = True to check for the vllm-tensorized marker.
+
+    # If `deserialization_params` is passed to TensorDeserializer as expected,
+    # we should expect the ValueError being caught.
```
Try something simpler like specifying an obviously invalid argument along the lines of `device="fish"`. This seems like a weird implementation bug to rely on here, and something that ought to be fixed anyway.

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -280,3 +292,196 @@ def test_vllm_tensorized_model_has_same_outputs(vllm_runner, tmp_path):
+    except RuntimeError:
+        out, err = capfd.readouterr()
+        combined_output = out + err
+        assert ("ValueError: Only binary modes (\"rb\", \"wb\", \"wb+\", etc.) are valid when opening local file streams.") in combined_output
```
You might not want to rely on the EXACT text of the error message here. That is not considered stable between tensorizer versions, and may change. Maybe just look for a `ValueError` in general.

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -280,3 +292,196 @@ def test_vllm_tensorized_model_has_same_outputs(vllm_runner, tmp_path):
+            f"{VLLM_PATH}/examples/others/tensorize_vllm_model.py", "--model",
+            model_ref, "serialize", "--serialized-directory",
+            str(tmp_path), "--suffix", suffix, "--serialization-kwargs",
+            json.dumps({"limit_cpu_concurrency": 4})
```
```suggestion
            '{"limit_cpu_concurrency": 4}'
```

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -280,3 +292,196 @@ def test_vllm_tensorized_model_has_same_outputs(vllm_runner, tmp_path):
+        print("STDERR:\n", e.stderr)
+        raise
+
+    print("Serializing STDOUT:\n", result.stdout)
```
I'm not sure if this is good to have here if it's not actually being used for a test assertion. Maybe just search for some simple substring or check that it's non-empty rather than printing it.

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -280,3 +292,196 @@ def test_vllm_tensorized_model_has_same_outputs(vllm_runner, tmp_path):
+        pytest.fail("Server did not start successfully")
+
+
+    proc.terminate()
```
Use `finally` to ensure the server is closed instead of letting it potentially hang indefinitely in a subprocess
```suggestion
    try:
        async with asyncio.timeout(180):
            await proc.stdout.readuntil(b"Application startup complete.")
    except asyncio.TimeoutError:
        pytest.fail("Server did not start successfully")
    finally:
        proc.terminate()
```

Eta0 commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -243,11 +256,15 @@ class TensorizerArgs:
-          raised if any of the hashes do not match.
+          raised if any of the hashes do not match. Deprecated, as this can 
+          now be passed along with any other `TensorDeserializer` kwargs 
+          through `deserialization_kwargs`.
```
Please clarify further that it's just passing it *here, specifically,* that is deprecated, lest people glean the impression that `verify_hash` is deprecated on the Tensorizer side.

Eta0 commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -243,11 +256,15 @@ class TensorizerArgs:
+          the number of readers based on the number of available.
           resources and model size. This greatly increases performance.
+          Deprecated, as this can now be passed along with any other 
+          `TensorDeserializer` kwargs through `deserialization_kwargs`.
```
Same comment as before regarding clarification on the deprecation.

Eta0 commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -258,6 +275,19 @@ class TensorizerArgs:
           S3_ENDPOINT_URL environment variable.
+      stream_kwargs: Arbitrary keyword arguments to pass to Tensorizer's
+          `stream_io` object, which the `TensorSeriaizer` and 
+          TensorDeserializer` objects use to work with streams.
```
1. The available keywords aren't really arbitrary—they are well-defined
2. `stream_io` is a module, not an object
3. Typos
```suggestion
      stream_kwargs: Keyword arguments to pass to Tensorizer's
          `stream_io.open_stream()` function, which the `TensorSerializer` and
          `TensorDeserializer` objects use to work with files and streams.
```

Eta0 commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -271,12 +301,9 @@ def __post_init__(self):
-            "verify_hash": self.verify_hash,
-            "encryption": self.encryption_keyfile,
-            "num_readers": self.num_readers
-        }
```
Please make sure that these three attributes are accounted for somewhere in `self.deserializer_kwargs` instead, even if they're deprecated.

sangstar replied:
Like I did with the handling of the stream params, I'll bring this code snippet back and have it be overwritten by `self.deserialization_kwargs` if available, like:

```python
        self.deserializer_params = {
            "verify_hash": self.verify_hash,
            "encryption": self.encryption_keyfile,
            "num_readers": self.num_readers
        }

        self.deserializer_params.update(**self.deserialization_kwargs)
```

Eta0 commented on vllm/model_executor/model_loader/tensorizer.py:
Relying purely on JSON to convey serializer and deserializer args only allows passing args with JSON-compatible types, e.g. strings, ints, bools, etc. For example, you can't currently pass a `dtype` for the deserializer like this, since that requires a `torch.dtype` object instead.
There isn't a great generic solution for this, but if e.g. specifying a `dtype` is something you'd expect people to want to do, we should at least special-case some extra parsing for those beyond `json.loads()`.

sangstar replied:
We provide Tensorizer the `dtype` downstream when deserializing with the `dtype` used for the LLM instance. It would be a bit weird to allow all arbitrary kwargs but then intervene if someone wants to specify the `dtype`, but I can't think of any good reason for the `dtype` provided for vLLM to be different to the `dtype` provided for Tensorizer, and someone passing both would just be confusing.

It might be worth checking for these JSON-incompatible before calling `json.loads` like `dtype`, but I wonder what to do if a different `dtype` is passed to, say, `vllm serve` and a different `dtype` to `TensorDeserializer`. It probably ought to raise a `ValueError`.

Eta0 replied:
There are a few parameters that would be pretty strange to override, yes. Maybe make a blacklist?

sangstar committed:
Update docs/models/extensions/tensorizer.md

Co-authored-by: Eta <24918963+Eta0@users.noreply.github.com>

sangstar committed:
Apply suggestions from code review

Apply the batch of commits suggested in this review.

Co-authored-by: Eta <24918963+Eta0@users.noreply.github.com>

sangstar committed:
fix: Properly assert type for serialization/deserialization args

Also fixes the logic for parsing different permutations for
using the example script based on whether args are passed to
CLI args directly or packaged in
--model-loader-extra-config

Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
tests: Resolve review comments on tests

Signed-off-by: Sanger Steel <sangersteel@gmail.com>

sangstar committed:
docs: Add clarifications on deprecated args, usage doc

Signed-off-by: Sanger Steel <sangersteel@gmail.com>

Eta0 commented on docs/models/extensions/tensorizer.md:
```diff
@@ -6,7 +6,7 @@ title: Loading models with CoreWeave's Tensorizer
-shorter Pod startup times and CPU memory usage. Tensor encryption is also supported.
+lower Pod startup times and CPU memory usage. Tensor encryption is also supported.
 
 vLLM fully integrates Tensorizer in to its model loading machinery. The
```
```suggestion
vLLM fully integrates Tensorizer in its model loading machinery. The
```

Eta0 commented on docs/models/extensions/tensorizer.md:
````diff
@@ -63,42 +65,50 @@ python examples/others/tensorize_vllm_model.py \
+one can simply do:
 
 ```bash
 #!/bin/bash
````
A shebang is excessive for a one line shell command.
```suggestion
```

Eta0 commented on docs/models/extensions/tensorizer.md:
```diff
@@ -63,42 +65,50 @@ python examples/others/tensorize_vllm_model.py \
+file locally to authenticate, but also can be given the environment variables 
+`AWS_ACCESS_KEY_ID` or `S3_ACCESS_KEY_ID` for the object storage access key,
+`AWS_SECRET_ACCESS_KEY` or `S3_SECRET_ACCESS_KEY` for the object storage 
+secret key, and `AWS_S3_ENDPOINT_URL` or `S3_ENDPOINT_URL` for the endpoint url.
```
1. Prefer to call it "object storage."
2. `AWS_S3_ENDPOINT_URL` is not the name of the environment variable (it's `AWS_ENDPOINT_URL_S3` or `AWS_ENDPOINT_URL`), but either way, we should just list the `AWS_*` ones since those are the only ones supported in Tensorizer itself, not any of the `S3_*` ones.
3. We should also note somewhere here that any normal [environment variables for configuring Boto3](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#using-environment-variables) will work.
```suggestion
Please note that object storage authentication is still required.
To authenticate, Tensorizer searches for a `~/.s3cfg` configuration file
([s3cmd](https://s3tools.org/kb/item14.htm)'s format),
or `~/.aws/config` and `~/.aws/credentials` files (`boto3` and the
`aws` CLI's format), but authentication can also be configured using
[any normal `boto3` environment variables](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html#using-environment-variables),
such as `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`,
and `AWS_ENDPOINT_URL_S3`.
```

Eta0 commented on docs/models/extensions/tensorizer.md:
````diff
@@ -63,42 +65,50 @@ python examples/others/tensorize_vllm_model.py \
 
-MODEL_LOADER_EXTRA_CONFIG=$(echo "$JSON" | tr -d '\n')
+```bash
+#!/bin/bash
````
Note that this is POSIX `sh`-compatible
```suggestion
#!/bin/sh
```

Eta0 commented on docs/models/extensions/tensorizer.md:
```diff
@@ -63,42 +65,50 @@ python examples/others/tensorize_vllm_model.py \
+If only the model tensors are saved, you can still provide that as the only 
+artifact in your directory to load from, and vLLM will fetch the rest using the 
+`--model-loader-extra-config` CLI arg, passing a JSON string with the args 
+for `TensorizerConfig`.
```
What does "vLLM will fetch the rest using the `--model-loader-extra-config` CLI arg" mean? Doesn't it fetch the rest using the positional argument for the model tag, not `--model-loader-extra-config`?

sangstar replied:
Yes, it expects the model tag to be a HuggingFace model ID. I'll make sure this is clear. I've done a poor job explaining that here.

Eta0 commented on examples/others/tensorize_vllm_model.py:
```diff
@@ -203,6 +204,14 @@ def parse_args():
+        required=False,
+        help=("A JSON string containing additional keyword arguments to "
+              "pass to Tensorizer's `TensorDeserializer` during "
+              "serialization."))
```
```suggestion
              "deserialization."))
```

Eta0 commented on examples/others/tensorize_vllm_model.py:
```diff
@@ -298,30 +307,36 @@ def deserialize():
-            tensorize_lora_adapter(args.lora_path, tensorizer_config)
+            if args.lora_path:
+                tensorizer_config.lora_dir = tensorizer_config.tensorizer_dir
+                tensorize_lora_adapter(args.lora_path, tensorizer_config)
```
It doesn't seem appropriate to ignore all of these if both they and `tensorizer_args` exist. If it is specified both ways, it should either simply raise an error telling you to pick one way, or merge the settings.

sangstar replied:
Good call. I'll have the `--model-loader-extra-config` JSON take precedence in updating values.

Eta0 commented on examples/others/tensorize_vllm_model.py:
```diff
@@ -298,30 +307,36 @@ def deserialize():
+                encryption_keyfile=keyfile,
+                **credentials
+            )
+            if args.serialization_kwargs:
+                serialization_kwargs = json.loads(args.serialization_kwargs)
```
Same comment I had before about this pattern and how it should be moved to the `argparse` type-parsing logic

sangstar replied:
It's being loaded by that now as per your commit.

Eta0 replied:
Why is it still here?

sangstar replied:
Will be removed in upcoming commit.

Eta0 commented on examples/others/tensorize_vllm_model.py:
```diff
@@ -298,30 +307,36 @@ def deserialize():
             )
+
+            if args.deserialization_kwargs:
+                deserialization_kwargs = json.loads(args.deserialization_kwargs)
```
(Again)
Same comment I had before about this pattern and how it should be moved to the `argparse` type-parsing logic

Eta0 commented on tests/tensorizer_loader/conftest.py:
```diff
@@ -7,11 +7,20 @@
 @pytest.fixture(autouse=True)
-def allow_insecure_serialization():
+def allow_insecure_serialization(monkeypatch):
     os.environ["VLLM_ALLOW_INSECURE_SERIALIZATION"] = "1"
```
You left this here
```suggestion
```

Eta0 commented on tests/tensorizer_loader/conftest.py:
```diff
@@ -29,3 +38,43 @@ def assert_from_collective_rpc(engine: LLM,
+        distributed_init_method = get_distributed_init_method(
+            get_ip(), get_open_port()
+        )
+        local_rank = 0
+        # set local rank as the device index if specified
+        device_info = self.vllm_config.device_config.device.__str__().split(
+            ":"
+        )
+        if len(device_info) > 1:
+            local_rank = int(device_info[1])
```
1. Why?
2. Some configurations will cause all processes to see their own separate GPUs as `cuda:0` despite having different local ranks. Can you just use typical ways of doing rank detection here rather than doing this?
3. If you do have to do this, call `str()`, don't invoke the `.__str__()` method.

sangstar replied:
As per previous conversations, this is just code from another test utils patched for this use case, so there's no real benefit to reviewing parts of the code not in use. I'll make clear what it's doing in the test script though.

Eta0 commented on tests/tensorizer_loader/conftest.py:
```diff
@@ -29,3 +38,43 @@ def assert_from_collective_rpc(engine: LLM,
+            distributed_init_method=distributed_init_method,
+            is_driver_worker=is_driver_worker,
+        )
+        self.collective_rpc("init_worker", args=([kwargs],))
```
A `dict` wrapped in a single element `list` wrapped in a single-element `tuple` sounds almost certainly incorrect, but if it is really supposed to be this, then oh no.

Eta0 commented on tests/tensorizer_loader/conftest.py:
```diff
@@ -29,3 +38,43 @@ def assert_from_collective_rpc(engine: LLM,
+        self.collective_rpc("init_worker", args=([kwargs],))
+        self.collective_rpc("init_device")
+
+    @property
+    def max_concurrent_batches(self) -> int:
+        return 2
```
Any reason not to do this?
```suggestion
    max_concurrent_batches = 2
```

sangstar replied:
As per prior discussions about this borrowed `DummyExecutor`, there's no need for changes to the dummy code itself that won't end up being used. I'll add a comment explaining clearly what it is and why it's there.

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -58,6 +68,22 @@
     os.path.dirname(__file__), "tensorize_vllm_model_for_testing.py")
 
 
+def patch_init_and_catch_error(self, obj, method, expected_error: Exception):
```
`expected_error` isn't an `Exception`, it's a type, so it should be annotated as `typing.Type[Exception]`

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -58,6 +68,22 @@
+
+    def wrapper(*args, **kwargs):
+        try:
+            breakpoint()
```
Stray breakpoint
```suggestion
```

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -357,29 +383,30 @@ def test_assert_deserialization_kwargs_passed_to_tensor_deserializer(tmp_path, c
     # lazy_load = True to check for the vllm-tensorized marker.
 
     # If `deserialization_params` is passed to TensorDeserializer as expected,
     # we should expect the ValueError being caught.
```
This comment is no longer accurate and should be removed.

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -357,29 +383,30 @@ def test_assert_deserialization_kwargs_passed_to_tensor_deserializer(tmp_path, c
+    vllm_config = engine_args.create_engine_config()
+    executor = DummyExecutor(vllm_config)
+
+    with pytest.raises(TensorizerCaughtError):
+        executor.collective_rpc(patch_init_and_catch_error, args=(tensorizer.serialization.TensorDeserializer, "__init__",  TypeError,))
```
You should leave a comment that you're translating `TypeError`s into `TensorizerCaughtError`s first instead of just checking `pytest.raises(TypeError)` directly to narrow the scope so that it only passes this test when `TensorDeserializer.__init__()` is the method throwing a `TypeError`.
At least, that's my reading of why you're doing that.

sangstar replied:
Good call, it looks confusing that I'm passing a `ValueError` in the function.

Eta0 commented on tests/tensorizer_loader/test_tensorizer.py:
```diff
@@ -402,26 +429,24 @@ def test_assert_stream_kwargs_passed_to_tensor_deserializer(tmp_path, capfd):
+    vllm_config = engine_args.create_engine_config()
+    executor = DummyExecutor(vllm_config)
+
+    import vllm.model_executor.model_loader.tensorizer
```
Move this import to the top of the file.

Eta0 commented on vllm/model_executor/model_loader/tensorizer.py:
```diff
@@ -61,6 +61,14 @@
 
 logger = init_logger(__name__)
 
+def tensorizer_kwargs_arg(value):
+    loaded = json.loads(value)
+    if not isinstance(loaded, dict):
+        raise argparse.ArgumentTypeError(
+            f"Not deserializable to dict: {value}. serialization_kwargs and "
+            f"deserialization_kwargs must be "
+            f"deserializable from a JSON string to a dictionary. "
+        )
```
This function is missing a return value...
If you were testing like this, it may have been switching the parsed value to `None` every time, so please test again with the corrected version.
```suggestion
def tensorizer_kwargs_arg(value):
    loaded = json.loads(value)
    if not isinstance(loaded, dict):
        raise argparse.ArgumentTypeError(
            f"Not deserializable to dict: {value}. serialization_kwargs and "
            f"deserialization_kwargs must be "
            f"deserializable from a JSON string to a dictionary. "
        )
    return loaded
```

Eta0 commented on tests/tensorizer_loader/conftest.py:
```diff
@@ -14,3 +31,50 @@ def cleanup():
+    return all(res)
+
+
+class DummyExecutor(UniProcExecutor):
```
Can you add a comment explaining why this class exists and what it does differently than others?

sangstar committed:
Apply suggestions from code review

Co-authored-by: Eta <24918963+Eta0@users.noreply.github.com>

sangstar committed:
Update docs/models/extensions/tensorizer.md

Co-authored-by: Eta <24918963+Eta0@users.noreply.github.com>

sangstar committed:
Update examples/others/tensorize_vllm_model.py

Co-authored-by: Eta <24918963+Eta0@users.noreply.github.com>

sangstar committed:
fix: Implement changes from second review

Signed-off-by: Sanger Steel <sangersteel@gmail.com>

Eta0 reviewed (approved):
All good. Just update the PR description with the simplified shell script, please.
